### PR TITLE
Implement jruby-puppet.environment-class-cache-enabled

### PR DESCRIPTION
**Pull request by jkroepke**
#454 

Its possible to add this to the foreman-installer?

### CODE REVIEW

File: templates/server/puppetserver/conf.d/puppetserver.conf.erb [hunk 1]
```diff
@@ -68,6 +68,11 @@ jruby-puppet: {
     # specified in the Puppet Server HOCON-formatted auth.conf (if false).
     use-legacy-auth-conf: <%= @server_use_legacy_auth_conf %>
 <%- end -%>
+
+<%- if scope.function_versioncmp([@server_puppetserver_version, '2.4']) >= 0 -%>
```
Comment (mmoll) [inline templates/server/puppetserver/conf.d/puppetserver.conf.erb]: the empty line should go below the if and not above, so it's looking nicer in any condition

**Comment by mmoll** [comment]
Maybe skip the "_enabled" for briefty - and adding a test with the value set would be nice. Adding the variable to the default params of the subclass in the existing test is needed anyway to make it pass.

File: manifests/server.pp [hunk 2]
```diff
@@ -363,100 +363,101 @@
...
+  $jvm_config                      = $::puppet::server_jvm_config,
+  $jvm_min_heap_size               = $::puppet::server_jvm_min_heap_size,
+  $jvm_max_heap_size               = $::puppet::server_jvm_max_heap_size,
+  $jvm_extra_args                  = $::puppet::server_jvm_extra_args,
+  $jruby_gem_home                  = $::puppet::server_jruby_gem_home,
+  $max_active_instances            = $::puppet::server_max_active_instances,
+  $max_requests_per_instance       = $::puppet::server_max_requests_per_instance,
+  $use_legacy_auth_conf            = $::puppet::server_use_legacy_auth_conf,
+  $check_for_updates               = $::puppet::server_check_for_updates,
+  $environment_class_cache_enabled = $::puppet::server_environment_class_cache_enabled,
```
Comment (domcleal) [inline manifests/server.pp]: The main `puppet` class needs this parameter adding - don't forget to add it to the docs at the top with `type:boolean`.

File: manifests/server.pp [hunk 3]
```diff
@@ -363,100 +363,101 @@
...
 ) {
 
   validate_bool($ca)
```
Comment (domcleal) [inline manifests/server.pp]: Also add a `validate_bool` for `$environment_class_cache_enabled` to the block around line 520 please.

**Comment by jkroepke** [comment]
I prefer that the variable name is exact the same as it should be write the puppetserver.conf.

> Adding the variable to the default params of the subclass in the existing test is needed anyway to make it pass.

It is already done, not?

File: manifests/server.pp (hunk 2 above)
Comment (jkroepke) [inline manifests/server.pp]: Done

File: manifests/server.pp (hunk 3 above)
Comment (jkroepke) [inline manifests/server.pp]: Done

**Comment by jkroepke** [comment]
Sorry for the confusion. Should be okay right now. (If the test passed)

File: manifests/init.pp [hunk 4]
```diff
@@ -4,606 +4,612 @@
...
 #
-# $dir_group::                        Group of the base puppet directory, used when
-#                                     puppet::server is false.
-#                                     type:string
+# $dir_group::                              Group of the base puppet directory, used when
+#                                           puppet::      server is false.
+#                                           type:string
 #
-# == puppet::server parameters
+# == puppet::      server parameters
```
Comment (domcleal) [inline manifests/init.pp]: Remove the whitespace here.

File: manifests/server.pp [hunk 5]
```diff
@@ -527,6 +528,8 @@
     validate_bool($ca_auth_required)
     validate_bool($use_legacy_auth_conf)
     validate_re($puppetserver_version, '^[\d]\.[\d]+\.[\d]+$')
+    validate_bool($environment_class_cache_enabled)
+
```
Comment (domcleal) [inline manifests/server.pp]: Linting is failing on this new line, remove it.

File: spec/classes/puppet_server_puppetserver_spec.rb [hunk 6]
```diff
@@ -28,38 +28,39 @@
...
+        :server_ca_client_whitelist             => [ 'localhost', 'puppetserver123.example.com' ],
+        :server_admin_api_whitelist             => [ 'localhost', 'puppetserver123.example.com' ],
+        :server_ruby_load_paths                 => [ '/some/path', ],
+        :server_ssl_protocols                   => [ 'TLSv1.2', ],
+        :server_cipher_suites                   => [ 'TLS_RSA_WITH_AES_256_CBC_SHA256',
+                                                     'TLS_RSA_WITH_AES_256_CBC_SHA',
+                                                     'TLS_RSA_WITH_AES_128_CBC_SHA256',
+                                                     'TLS_RSA_WITH_AES_128_CBC_SHA', ],
+        :server_max_active_instances            => 2,
+        :server_max_requests_per_instance            => 0,
```
Comment (mmoll) [inline spec/classes/puppet_server_puppetserver_spec.rb]: indentation looks off here

File: spec/classes/puppet_server_puppetserver_spec.rb [hunk 7]
```diff
@@ -28,38 +28,39 @@
...
+        :server_puppetserver_rundir             => '/var/run/puppetlabs/puppetserver',
+        :server_puppetserver_logdir             => '/var/log/puppetlabs/puppetserver',
+        :server_jruby_gem_home                  => '/opt/puppetlabs/server/data/puppetserver/jruby-gems',
+        :server_dir                             => '/etc/puppetlabs/puppet',
+        :codedir                                => '/etc/puppetlabs/code',
+        :server_idle_timeout                    => 1200000,
+        :server_connect_timeout                 => 120000,
+        :server_enable_ruby_profiler            => false,
+        :server_check_for_updates               => true,
+        :server_environment_class_cache_enabled => true,
```
Comment (mmoll) [inline spec/classes/puppet_server_puppetserver_spec.rb]: if the default false, it should also be false here.

File: templates/server/puppetserver/conf.d/puppetserver.conf.erb [hunk 8]
```diff
@@ -67,6 +67,11 @@ jruby-puppet: {
     # in the legacy Puppet auth.conf file (if true or not specified) or via rules
     # specified in the Puppet Server HOCON-formatted auth.conf (if false).
     use-legacy-auth-conf: <%= @server_use_legacy_auth_conf %>
+
```
Comment (mmoll) [inline templates/server/puppetserver/conf.d/puppetserver.conf.erb]: now that empty line will always be there, it should be roughly:
```
use-legacy-auth-conf: <%= @server_use_legacy_auth_conf %>
<%- end -%>
<%- if scope.function_versioncmp([@server_puppetserver_version, '2.4']) >= 0 -%>

 # (optional) enable or disable environment class cache
```

**Comment by mmoll** [comment]
we're getting there :+1:

**Comment by jkroepke** [comment]
I have no idea, why the tests failed. Can someone help? I have idea why the default is true while false is defined in params.

**Comment by mmoll** [comment]
@jkroepke https://github.com/theforeman/puppet-puppet/pull/455/files#r88004813

**Comment by domcleal** [comment]
See https://github.com/theforeman/puppet-puppet/pull/455/files#r88004813, default_params in the test.

**Comment by jkroepke** [comment]
Damn. Thanks.

File: templates/server/puppetserver/conf.d/puppetserver.conf.erb (hunk 8 above)
Comment (mmoll) [inline templates/server/puppetserver/conf.d/puppetserver.conf.erb]: @jkroepke could you please change the empty line?

File: manifests/init.pp [hunk 9]
```diff
@@ -4,606 +4,615 @@
...
-#
-# $server_check_for_updates::         Should the puppetserver phone home to check for available updates?
-#                                     Defaults to true
-#                                     type:boolean
+# $server_environment_class_cache_enabled:: Enable environment class cache
+#                                           Note: If you’re not using Code Manager and opt to enable this setting, make
+#                                           sure your code deployment method – for example r10k – clears the environmen
+#                                           cache when it completes. If you don’t clear the environment cache, the Node
+#                                           Classifier doesn’t receive new class information until Puppet server is
+#                                           restarted.
```
Comment (mmoll) [inline manifests/init.pp]: I'm unsure if that note text should be really placed here as is, espcially when thinkg about the resulting long output in the foreman-instalelr help and interactive mode. There are many consequences of more or less any switch we offer, still we don't warn at each of these.

**Comment by mmoll** [comment]
@jkroepke ping?

**Comment by jkroepke** [comment]
I have no idea, which description is in your mindset. Feel free to take better suggestions. I don't have the skills the write a own good description, since my english is limited.

**Comment by mmoll** [comment]
I pushed the changes I had in mind and am OK with it to be squashed into one commit.

**Comment by domcleal** [comment]
Thanks @jkroepke!

**Pull request merged**